feature: history stack gui initial version

SeolJaeHyeok opened the pull request:
## Summary

### 주요 컴포넌트 설명

#### 1. HistoryManager(src/devtools/HistoryManager.ts)
- History API 래핑 및 상태 관리
- observer pattern을 통한 이벤트 구독
- 내부, 외부 링크 구분 및 처리

#### 2. HistoryDevTools(src/devtools/HistoryDevTools.tsx)
- 히스토리 스택 시각화 UI
- 실시간 업데이트 및 인터랙션 처리

### 사용 방법

```typescript
import HistoryDevTools from './devtools/HistoryDevTools';

// 앱의 최상위 컴포넌트에 추가
<HistoryDevTools 
  initialIsOpen={true}
  panelPosition="bottom-right"
  buttonPosition="bottom-right"
/>
```

## Visuals

https://github.com/user-attachments/assets/6c3acb75-fb40-4203-9379-38277fda0aa8

## Note

- 현재 테스트 코드가 존재하지 않습니다. 당장은 유닛테스트까지만 하면 되지 않을까 하는데... 테스트에 대한 경험이 부족해서 뭘 이용해서 어떤 범위까지 해야 하는지 감이 잘 안오네요😂 좋은 의견 있으면 부탁드립니다:)

SeolJaeHyeok committed:
create vite project using pnpm

SeolJaeHyeok committed:
add: prettier dependency

SeolJaeHyeok committed:
fix: tsconfig compilerOptions

SeolJaeHyeok committed:
feature: History Devtools 기본 구조 생성
1. UI 구조 잡기

SeolJaeHyeok committed:
feature: History Manager Class 추가
1. 옵저버 패턴 활용
2. 기존 history api를 오버라이드하고 클래스 내부 상태로 현재 history에 대한 정보를 저장

SeolJaeHyeok committed:
chore: prettier configuration 파일 확장자 수정 및 전체 적용

SeolJaeHyeok committed:
feature: 테스트 컴포넌트 추가
1. 테스트 페이지를 구성하고 각 페이지 내에 다른 페이지로 이동할 수 있는 버튼 추가
2. 이를 위해 react router dom 의존성 추가

SeolJaeHyeok committed:
feature: a tag로 이동하는 경우 대응
1. a 태그를 가로채셔 기본동작을 제어
2. 특수한 경우(다운로드, 외부 링크 등)는 기본 동작을 유지

SeolJaeHyeok committed:
fix: 스택 존재 유무 판별 로직 수정

SeolJaeHyeok committed:
feature: a 태그의 이동인 경우에도 react router를 사용하는 경우와 동일한 state를 만들어서 주입하기

SeolJaeHyeok committed:
feature: devtools position props

hovelopin commented on milkboy/eslint.config.js:
```diff
@@ -0,0 +1,28 @@
+import globals from 'globals'
+import reactHooks from 'eslint-plugin-react-hooks'
+import reactRefresh from 'eslint-plugin-react-refresh'
+import tseslint from 'typescript-eslint'
```
오,, typescript-eslint를 쓰는건 처음 보는데 기본 eslint에서 typescript 관련 plugin을 설치했을떄와 동일한 역할을 하는건가요??

hovelopin replied:
아 요거 vite 기본 설정이군요!

SeolJaeHyeok replied:
맞습니다!
기본 설정 외의 별도의 설정을 추가하지는 않았아요:)

hovelopin commented on milkboy/src/App.tsx:
```diff
@@ -0,0 +1,19 @@
+
+  return (
+    <>
+      <button onClick={() => navigate('/test1')}>test1</button>
+      <button onClick={() => navigate('/test2')}>test2</button>
+      <button onClick={() => navigate('/test3')}>test3</button>
+      <button onClick={() => navigate('/test4')}>test4</button>
+      <button onClick={() => navigate(-1)}>뒤로가기</button>
+      <button onClick={() => navigate(1)}>앞으로가기</button>
```
테스트 해봤는데 replace 옵션도 Stack 변화에 따라 잘 동작하는 것 같더라고요! 해당 예제도 추가가되면 좋을 것 같습니다.

SeolJaeHyeok replied:
오 좋은 의견 감사합니다👍

hovelopin commented on milkboy/src/devtools/HistoryManager.ts:
```diff
@@ -0,0 +1,231 @@
+  search: string; // URL의 쿼리 파라미터
+  hash: string; // URL의 해시
+  state: unknown; // history.state에 저장된 상태
+  timestamp: number; // 항목이 생성된 시간
```
timestamp의 용도는 어떻게 될까요??

SeolJaeHyeok replied:
timestamp는 해당 라우트에 접근한 시간을 나타내기 위함입니다!(로깅 용도)

해당 필드를 추가하면서도`과연 이 값을 유의미하게 사용할 수 있을까`에 대한 막연한 고민이 있었는데, 
혹시 조금 더 디테일한 상황(e.g. 해당 페이지의 체류 시간 측정..?)에서 사용할 수 있지 않을까 했습니다! 

<img width="311" alt="image" src="https://github.com/user-attachments/assets/5f0c9c0d-3fff-458b-96b3-151bb5cec5bb" />

hovelopin replied:
넵! 이해했습니다.

hovelopin commented on milkboy/src/devtools/HistoryManager.ts:
```diff
@@ -0,0 +1,231 @@
+    try {
+      // 현재 도메인과 다른 외부 링크 체크
+      const url = new URL(href, window.location.origin);
+      if (url.origin !== window.location.origin) return false;
+
+      return true;
```
```js
return url.origin === window.location.origin
```

이거 별건 아닌데 같은지 여부 판단해서 같으면 true 다르면 false 로직이라면 간단하게 처리도 가능할 것 같습니다잉~

SeolJaeHyeok replied:
말씀하신대로 작성하는게 조금 더 간단하게 처리하는 좋은 방식인거 같아용!

근데 저는 조건이 여러개가 되면 해당 방식이 조금 더 가독성을 해치는 상황이 발생하는 경험이 많았어요..!

예를 들어, 다음과 같은 상황이었습니다.

```javascript
return a === b && c !== d && e === true
```

위 방식이 코드 파악할 때 인지적인 부담을 증가시킨다고 생각해서, 저는 아래와 같이 Early Return을 활용해서 작성하는걸 선호하는 편입니다!

```javascript
if(a !== b) {
  return false;
};

if(c === d) {
  return false;
}

if(e === false) {
  return false;
}

return true;
```

이렇게 하면 위에서부터 하나씩 훑어가면서 조건을 파악할 수 있어서 개인적으로는 읽기 더 편하더라고요!
추가로, 주석을 활용할 때에도 이 방식이 조금 더 낫지 않나 하는 생각이었어요.
물론 코드의 길이가 이전보다 길어진다는 문제는 있지만 이게 단점이라고는 생각하지 않았어요!

그래서 조건이 하나인 경우에도 지금과 같이 적어줬습니다~

hovelopin replied:
재혁님이 말씀해주신대로 저도 조건이 길어지면 `early-return`을 선호하는데 해당 케이스는 1가지 조건만 다루기 때문에 이해하는것에는 문제가 없다고 생각했었습니다! 

로직에는 크게 문제가 없고 코드 스타일의 문제일 것 같아서 일단은 현재 작성되어 있는대로 진행하시죠 👍👍

hovelopin commented on milkboy/src/devtools/HistoryManager.ts:
```diff
@@ -0,0 +1,231 @@
+      if (url.origin !== window.location.origin) return false;
+
+      return true;
+    } catch {
+      // 상대 경로인 경우 내부 링크로 간주
+      return true;
```
그리고 이 부분이 이해가 안가는데 new URL 생성자에서 에러를 반환하면 catch에 잡히는 것 같은데 이게 왜 true가 되어야하는지 이해가 안갑니다!

SeolJaeHyeok replied:
이거 저도 확인해봤는데, 
처음 의도는 `[slug]`, `posts/[id]`, `(root)` 와 같은 nextjs의 동적 라우팅을 대응하려고 작성을 했었는데, 
테스트해보니 해당 경우에 정상적인 경로로 처리가 되서 catch 문으로 이동하는게 아니더라구요!

우선은 try/catch가 필요하지 않을거 같아서 제거했습니다.
추후에 테스트 코드를 도입하면서 조금 더 고려해봐야겠네요!

좋은 의견 감사합니다!

hovelopin reviewed (commented):
간단하게 데모버전만 작업한 것 같은데 컨셉이 너무 좋은 것 같습니다! 컨셉 유지하고 계속 진행해보시죠!

SeolJaeHyeok committed:
feature: add replace example

SeolJaeHyeok commented:
> 간단하게 데모버전만 작업한 것 같은데 컨셉이 너무 좋은 것 같습니다! 컨셉 유지하고 계속 진행해보시죠!

아 그리고 추가적으로 react-router는 `popstate` 이벤트를 감지해서 라우팅을 처리하게 되는데 `history.pushState`만 사용하면 react-router가 라우팅 변경을 감지하지 못해서 렌더링을 발생시키지 않는 문제가 있었습니다!

따라서 a 태그를 클릭한 경우 `popstate` 이벤트를 임의로 발생시켜주는 로직이 추가되었습니다. 

```javascript
const popStateEvent = new PopStateEvent('popstate', { state: newState });
window.dispatchEvent(popStateEvent);
```

참고 부탁드립니다!

SeolJaeHyeok committed:
fix: additional task
1. remove console.log
2. remove useless try/catch
3. add popstate event for react-router
